components/script/dom/bindings/like.rs

```diff
@@ -0,0 +1,301 @@
+
+        #[inline(always)]
+        fn add(&$self, key: Self::Key) {
+            $self.$x.add(key)
+        }
+
+        #[inline(always)]
+        fn has(&$self, key: Self::Key) -> bool {
+            $self.$x.has(key)
+        }
+
+        #[inline(always)]
+        fn clear(&$self) {
+            $self.$x.clear()
+        }
+
+        #[inline(always)]
+        fn delete(&$self, key: Self::Key) -> bool {
+            $self.$x.delete(key)
+        }
+    };
+}
+
+/// Every Maplike dom_struct must implement this
+/// to provide access to underlying storage
+/// so codegen can automatically generate all maplike methods
+///
+/// In case you use type that implements Maplike as underlying storage it's recommended to use `maplike` macro.
+// This is somehow analogues to gecko's MaplikeHelpers
+pub trait Maplike {
+    // maplike<Key, Value>
+    /// The type of the key of the map.
+    type Key: ToJSValConvertible;
+    /// The type of the value of the map.
+    type Value: ToJSValConvertible;
+
+    fn get_index(&self, index: u32) -> Option<(Self::Key, Self::Value)>;
+
+    fn get(&self, key: Self::Key) -> Option<Self::Value>;
+    fn size(&self) -> u32;
+    fn set(&self, key: Self::Key, value: Self::Value);
+    fn has(&self, key: Self::Key) -> bool;
+    fn clear(&self);
+    fn delete(&self, key: Self::Key) -> bool;
+}
+
+impl<T: Maplike> Iterable for T {
+    type Key = T::Key;
+
+    type Value = T::Value;
+
+    #[inline]
+    fn get_iterable_length(&self) -> u32 {
+        self.size()
+    }
+
+    #[inline]
+    fn get_value_at_index(&self, index: u32) -> Self::Value {
+        // SAFETY: we are checking bounds manually
+        self.get_index(index).unwrap().1
+    }
+
+    #[inline]
+    fn get_key_at_index(&self, index: u32) -> Self::Key {
+        // SAFETY: we are checking bounds manually
+        self.get_index(index).unwrap().0
+    }
+}
+
+impl<K, V> Maplike for DomRefCell<IndexMap<K, V>>
+where
+    K: ToJSValConvertible + Eq + PartialEq + Hash + Clone,
+    V: ToJSValConvertible + Clone,
+{
+    type Key = K;
+    type Value = V;
+
+    #[inline(always)]
+    fn get_index(&self, index: u32) -> Option<(Self::Key, Self::Value)> {
+        self.borrow()
+            .get_index(index as usize)
+            .map(|(k, v)| (k.to_owned(), v.to_owned()))
+    }
+
+    #[inline(always)]
+    fn get(&self, key: Self::Key) -> Option<Self::Value> {
+        self.borrow().get(&key).cloned()
+    }
+
+    #[inline(always)]
+    fn size(&self) -> u32 {
+        self.borrow().len() as u32
+    }
+
+    #[inline(always)]
+    fn set(&self, key: Self::Key, value: Self::Value) {
+        self.borrow_mut().insert(key, value);
+    }
+
+    #[inline(always)]
+    fn has(&self, key: Self::Key) -> bool {
+        self.borrow().contains_key(&key)
+    }
+
+    #[inline(always)]
+    fn clear(&self) {
+        self.borrow_mut().clear()
+    }
+
+    #[inline(always)]
+    fn delete(&self, key: Self::Key) -> bool {
+        // TODO: shift_remove ???
```

Comment:
Is this comment still valid?

Reply:
I rechecked the specs and am pretty sure we need to keep order: https://tc39.es/ecma262/multipage/keyed-collections.html#sec-set.prototype.delete, so shift_remove is more appropriate.